### ISSUE TITLE
CSV Import of `sku,attribute` empties `url_key` value

### ISSUE THREAD

**Issue by josephmcdermott** [magento/magento2#17023]
### Preconditions
1. Magento 2.2.5 installation with a product created.

### Steps to reproduce
1. Import a partial CSV with just `sku,description` columns.
2. Check the Product to find its URL Key is gone.

### Expected result
1. URL Key should not be altered if it wasn't included in the CSV.

### Actual result
1. URL Key is changed to empty.
2. See \Magento\CatalogImportExport\Model\Import\Product::getUrlKey

**Comment by magento-engcom-team** [comment]
Hi @josephmcdermott. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me {$VERSION} instance` 

where `{$VERSION}` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (2.2-develop +). 
For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation.

@josephmcdermott do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by ghost** [comment]
@josephmcdermott, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by valenciaisaza** [comment]
Hi

I have the same problem with the same magento verison 2.2.5, I've been debugging the code to figure it out what happends, because in another magento 2.2.2 I haven't this issues . I found a difference on a validation on the file vendor/magento/module-catalog-import-export/Model/Import/Product/Type/AbstractType.php (Line 537). And this change make the url_key dissapear.

Line on magento 2.2.2 (works properly)
            if (!$attrParams['is_static'] && empty($rowData[$attrCode])) {

same line on Magento 2.2.4 and 2.2.5 (Fails)
            if (!$attrParams['is_static'] && !isset($rowData[$attrCode])) {


I hope it helps to someone.

**Comment by gabriel3iri** [comment]
Hello @magento-engcom-team , do you know why this validation was changed?

**Comment by chris-pook** [comment]
The issue looks to me to be simply related to the handling of the url_key column in this function called in the main loop for each row of the product import sheet:

```
foreach ($bunch as $rowNum => $rowData) {
    ...
    $rowData[self::URL_KEY] = $this->getUrlKey($rowData);
    ...
}
```

```
    protected function getUrlKey($rowData)
    {
        if (!empty($rowData[self::URL_KEY])) {
            return strtolower($rowData[self::URL_KEY]);
        }

        if (!empty($rowData[self::COL_NAME])) {
            return $this->productUrl->formatUrlKey($rowData[self::COL_NAME]);
        }

        return '';
    }
```

If the url_key column is not set in the data, then it will be set to '', and subsequently clears out the product's url attribute. I think there needs to be a tighter check of if the import sheet is intending to update the url_key or not here.

**Comment by evilprophet** [comment]
I also found other fid for this issue. 
File: `Magento\CatalogImportExport\Model\Import\Product`
Method: `_saveProducts`
After change line: 
```
$rowData[self::URL_KEY] = $urlKey;
```
to:
```
$urlKey = $this->getUrlKey($rowData);
if (!empty($urlKey)) {
    $rowData[self::URL_KEY] = $urlKey;
}
```
then column won't be set in import data, and url_key won't be set to empty.

**Comment by josephmcdermott** [comment]
Unfortunately that's not quite enough @evilprophet as if you import a CSV with `sku,name` only, all of the URL Keys will be replaced with the value provided in the name column.

I'm having a look at this currently, but have to say I'm disappointed at the lack of input from the Magento core team @magento-engcom-team since this is a pretty major issue.

**Comment by josephmcdermott** [comment]
@evilprophet take a look at the above PR, I think that covers all scenarios that I can tell. Of course it would be great to get feedback from @magento-engcom-team also to perform their own tests.
- PR: https://github.com/magento/magento2/pull/17882

**Comment by evilprophet** [comment]
@josephmcdermott Yes, you are right I forgot about case when there is a `name` without `url_key`. Your solution looks a bit better.

**Comment by erikhansen** [comment]
For the record, I applied PR #17882 as a composer patch to a 2.2.5 Commerce installation and it fixed this issue for me. The merchant had a custom text field attribute and the url_key value was getting deleted when they uploaded a CSV that contained that column. Thanks for your contribution, @josephmcdermott!

**Comment by magento-engcom-team** [comment]
Hi @josephmcdermott. Thank you for your report. 
The issue has been fixed in magento/magento2#17882 by @josephmcdermott in 2.2-develop branch
Related commit(s):
 - [a021d5f72b2c23be2820225d869b72a48d437530](https://github.com/magento/magento2/commit/a021d5f72b2c23be2820225d869b72a48d437530)
 - [a02bf03e26c3409a0813e4dc4e74aa8ccf21eb51](https://github.com/magento/magento2/commit/a02bf03e26c3409a0813e4dc4e74aa8ccf21eb51)
 - [e48683cd6827ff527a7f22f6c2e39a8a2922bd9a](https://github.com/magento/magento2/commit/e48683cd6827ff527a7f22f6c2e39a8a2922bd9a)
 - [7df4e129dd40e1659e4f81554d1a8ba37e6850f8](https://github.com/magento/magento2/commit/7df4e129dd40e1659e4f81554d1a8ba37e6850f8)
 - [fffb6ff6e8b4302db2f1b824fbe96a5cdacd90ff](https://github.com/magento/magento2/commit/fffb6ff6e8b4302db2f1b824fbe96a5cdacd90ff)
 - [002aa19711634a906aab53d6e1110babf604892e](https://github.com/magento/magento2/commit/002aa19711634a906aab53d6e1110babf604892e)
 - [172e2ec1ffd24e04e0050e58b7cb090785ebd9a1](https://github.com/magento/magento2/commit/172e2ec1ffd24e04e0050e58b7cb090785ebd9a1)
 - [3a54946f843d469b441185a36fc1a986e7e941fc](https://github.com/magento/magento2/commit/3a54946f843d469b441185a36fc1a986e7e941fc)
 - [2fb71834f2d0ce7282afda3d350cf404ad35be8f](https://github.com/magento/magento2/commit/2fb71834f2d0ce7282afda3d350cf404ad35be8f)
 - [fda664a268303561cd3acfbb35b52f85ae525ce5](https://github.com/magento/magento2/commit/fda664a268303561cd3acfbb35b52f85ae525ce5)
 - [abd95bdc0ab572881f8b341902d7115900fd0e39](https://github.com/magento/magento2/commit/abd95bdc0ab572881f8b341902d7115900fd0e39)
 - [0f26b52735389f412eb883eec978cd12052e0d97](https://github.com/magento/magento2/commit/0f26b52735389f412eb883eec978cd12052e0d97)

The fix will be available with the upcoming 2.2.8 release.
